test/functional/p2p_leak.py

```diff
@@ -106,17 +94,24 @@ def set_test_params(self):
         self.num_nodes = 1
 
     def run_test(self):
+        # Peer that never sends a version. We will send a bunch of messages
+        # from this node anyway and verify eventual disconnection.
         no_version_disconnect_node = self.nodes[0].add_p2p_connection(
-            CNodeNoVersionMisbehavior(), send_version=False, wait_for_verack=False)
-        no_version_idlenode = self.nodes[0].add_p2p_connection(CNodeNoVersionIdle(), send_version=False, wait_for_verack=False)
+            CLazyNode(), send_version=False, wait_for_verack=False)
+
+        # Another peer that never sends a version. Just sits idle and hopes to receive
+        # any message (it shouldn't!)
+        no_version_idlenode = self.nodes[0].add_p2p_connection(CLazyNode(), send_version=False, wait_for_verack=False)
+
+        # Peer that sends a version but not a verack.
         no_verack_idlenode = self.nodes[0].add_p2p_connection(CNodeNoVerackIdle(), wait_for_verack=False)
 
-        # Send enough veracks without a message to reach the peer discouragement
-        # threshold. This should get us disconnected.
+        # Send enough ping messages (any non-version message will do) prior to sending
+        # version to reach the peer discouragement threshold. This should get us disconnected.
         for _ in range(DISCOURAGEMENT_THRESHOLD):
-            no_version_disconnect_node.send_message(msg_verack())
+            no_version_disconnect_node.send_message(msg_ping())
 
```

Comment:
Great suggestion - I did my best but please let me know if there's a better way

Reply:
@gzhao408 After I added this assertion, the test sporadically failed. You and I discussed it and thought it was perhaps the timeout. But increasing it didn't seem to solve the problem. @amitiuttarwar gave me hints [here](https://github.com/bitcoin/bitcoin/pull/19727#pullrequestreview-468852817). It turns out that occasionally, the peer has been dropped prior to the test making `getpeerinfo` RPC. And at that point, the misbehaving peer isn't even in the peer list.

Since this particular peer does not do anything else (not even a version call) except overwhelm the node with pings, I thing the assertion that makes sure the connection is dropped suffices. Would you agree?